### PR TITLE
Fix/upsert graph 782

### PR DESCRIPTION
**Pull request by johannessjoberg**
This PR is aimed to solve this issue presented here: https://github.com/Vincit/objection.js/issues/782

In this discussion @koskimas presented 3 criteria to merge a potential PR: 

- [x] All current tests must pass (this should be obvious)

- [x] The new feature must be tested

- Integration tests are added for verifying that relating node with children is working for `ManyToMany`, `BelongsToOne` and `HasMany` relations
- Unit tests are added to verify that options object is correctly transformed depending on path from root
- Integration test for verifying that options are respected when relating node with children 
- [x] Must support all current options with boolean and relation path values. Especially remember to consider the case where relation paths point to models that are under the model that is about to be related.
- See `transformOptionsFromPath` function and unit tests

### CODE REVIEW

File: lib/queryBuilder/operations/UpsertGraphOperation.js [hunk 1]
```diff
@@ -176,6 +178,23 @@ class UpsertGraphOperation extends InsertGraphOperation {
...
+
+    return Promise.map(
+      upsertRecursivelyNodes,
+      node => {
+        return node.upsertModel.constructor
+          .query()
+          .upsertGraph(node.upsertModel, transformOptionsFromPath(node.opt, node.relPathFromRoot))
+          .childQueryOf(builder, true);
+      },
+      { concurrency: this.graph.rootModelClass.concurrency }
```
Comment (igor-savin-ht) [inline lib/queryBuilder/operations/UpsertGraphOperation.js]: Considering that knex is intending to eventually drop Bluebird usage, is it safe relying on it being there? @koskimas

File: lib/utils/transformOptionsFromPath.js [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    const transformedOpt = transformOptionFromPath(options[key], relPathFromRoot);
+    if (transformedOpt) {
+      transformedOptions[key] = transformedOpt;
+    }
+  });
+
+  return transformedOptions;
+};
+
+const transformOptionFromPath = (opt, relPathFromRoot) => {
```
Comment (igor-savin-ht) [inline lib/utils/transformOptionsFromPath.js]: since this is a public function, probably having jsdoc would be a good idea

Comment (koskimas) [reply lib/utils/transformOptionsFromPath.js]: ? This is not public

Comment (koskimas) [reply lib/utils/transformOptionsFromPath.js]: And jsdoc isn't used anywhere in objection codebase

Comment (igor-savin-ht) [reply lib/utils/transformOptionsFromPath.js]: @koskimas Right, my mistake. Comment still applies, but should be at "transformOptionsFromPath" instead.

Comment (koskimas) [reply lib/utils/transformOptionsFromPath.js]: That's not public either. Only stuff mentioned in the docs is considered public

File: lib/utils/transformOptionsFromPath.js [hunk 3]
```diff
@@ -0,0 +1,32 @@
+const transformOptionsFromPath = (options, relPathFromRoot) => {
```
Comment (igor-savin-ht) [inline lib/utils/transformOptionsFromPath.js]: Probably would be a good idea to add jsdoc since this is a public function. Also, any particular reason why this is defined as a constant arrow function? In other parts of objection.js I see exported functions defined simply as "function xyz (args)"

File: lib/utils/transformOptionsFromPath.js (hunk 2 above)
Comment (igor-savin-ht) [reply lib/utils/transformOptionsFromPath.js]: Fair point.

**Comment by koskimas** [comment]
Oh, It's not travis' problem. Your tests fail for some reason. Objection simply swallows the actual error and shows the useless one seen in travis :grimacing: There is something wrong with your migrations I'd quess. I just tested and the master's test run without any problem.

File: lib/utils/transformOptionsFromPath.js (hunk 3 above)
Comment (koskimas) [reply lib/utils/transformOptionsFromPath.js]: I agree, the same style could be used for these for consistency

Comment (johannessjoberg) [reply lib/utils/transformOptionsFromPath.js]: All right, got some different habits :) I'll fix that. Perhaps directly export the function?

**Comment by johannessjoberg** [comment]
Thanks @koskimas. I'll get on that rebase and try to figure out what's wrong.

File: lib/queryBuilder/operations/UpsertGraphOperation.js (hunk 1 above)
Comment (johannessjoberg) [reply lib/queryBuilder/operations/UpsertGraphOperation.js]: 👍 It's also used in similar fashion for `delete`, `relate` and `update` operation currently.

**Comment by johannessjoberg** [comment]
If seems as if travis does like me like me much :| 

Builds pass for Node 9 prev commit, this commit for Node 6. But for others Node versions: `No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.`

**Comment by koskimas** [comment]
@johannessjoberg The last one was just a random travis glitch. I restarted the tests and they finished successfully

**Comment by igor-savin-ht** [comment]
@koskimas What is left before this can be merged?

**Comment by koskimas** [comment]
@igor-savin-ht I haven't reviewed the code yet.

File: lib/queryBuilder/graphUpserter/UpsertNode.js [hunk 4]
```diff
@@ -327,6 +348,22 @@ function hasChanges(currentModel, upsertModel) {
...
+function hasRelationsInUpsertModel(upsertModel) {
+  const relationArray = upsertModel.constructor.getRelationArray();
+  const changingProps = [];
+
+  for (let i = 0, l = relationArray.length; i < l; ++i) {
+    const relation = relationArray[i];
+    const upsertRelated = upsertModel[relation.name];
+
+    if (upsertRelated) {
+      changingProps.push(upsertRelated);
```
Comment (koskimas) [inline lib/queryBuilder/graphUpserter/UpsertNode.js]: The array is not needed. You could just return true here and remove the array right?

File: testUtils/TestSession.js [hunk 5]
```diff
@@ -189,26 +216,78 @@ class TestSession {
...
+              .index();
+            table
+              .integer('model2Id')
+              .unsigned()
+              .notNullable()
+              .references('id_col')
+              .inTable('model2')
+              .onDelete('CASCADE')
+              .index();
           })
```
Comment (koskimas) [inline testUtils/TestSession.js]: Why did you change all these? Do you have different settings for `prettier` or something?

Comment (koskimas) [reply testUtils/TestSession.js]: Oh, I think my prettier script doesn't touch this file at all. This is fine :+1:

**Comment by koskimas** [comment]
@johannessjoberg Everything looks good to me. I added a note about a small improvement (one array creation could be avoided). Other than that I'm ready to merge this. Awesome job, excellent tests, clean code :100:

**Comment by koskimas** [comment]
@johannessjoberg Actually the one nitpick I found is so small that I don't want to bother you with it. Thanks again!

**Comment by johannessjoberg** [comment]
Glad to hear everything looks good and thank you for the kind words! Always glad to be able to help out :) 

When do you think this will be released?

**Comment by koskimas** [comment]
Soon, maybe even today.

**Comment by johannessjoberg** [comment]
Awesome! 👍

**Comment by koskimas** [comment]
@johannessjoberg 1.1.9 is out including this PR

**Comment by johannessjoberg** [comment]
Great! Thanks @koskimas :)
